### ISSUE TITLE
ONOS transaction stuck in COMMITTED state after redeployment with persistent volume

### ISSUE THREAD

**Issue by amitinfo2k** [onosproject/onos-config#1626]
### Issue description
We have deployed ROC using the latest AIAB release 2.1 (CHARTS=release-2.1 make roc-4g-models) while deployment we have enabled PV for consensus store. We did some operation where we had to redeploy the aether roc where we  cleaned up the roc using make roc-clean and then delete the PV. 
Then we redeployed the roc with the VP backing the same directory.So, it recovered the old configuration but, when we tried to perform new transactions e.g. updating the subscribers, enterprise etc. The new transaction keeps in COMMITED; it doesn't get APPLIED.
We are seeing the following error on onos-consensus-store as well as onos-config-atomix-proxy.

Note : This is not consistent and may be reproduced in 2-3 attempts of redeployment.

### Error in onos-consensus-store log:
`2023-02-09T06:08:49.285Z        WARN    [github.com/atomix/runtime/primitives/pkg/indexedmap/v1](http://github.com/atomix/runtime/primitives/pkg/indexedmap/v1)  v1/server.go:354        Events  {"EventsRequest": "headers:<operation:<primitive:<session:<partition:<partition_id:1 > session_id:7416 > primitive_id:2 > > sequence_num:45 > input:<key:\"uuid:310d6b62-2a88-419e-acc0-f2121eccff5e\" > ", "error": "rpc error: code = Canceled desc = context canceled"}
2023-02-09T09:33:32.884Z        WARN    [github.com/atomix/runtime/primitives/pkg/indexedmap/v1](http://github.com/atomix/runtime/primitives/pkg/indexedmap/v1)  v1/server.go:354        Events  {"EventsRequest": "headers:<operation:<primitive:<session:<partition:<partition_id:1 > session_id:7416 > primitive_id:2 > > sequence_num:106 > input:<key:\"uuid:d177e9e1-a403-4c4f-bf47-4e7a01c136f3\" > ", "error": "rpc error: code = Canceled desc = context canceled"}
`
### Error in onos-config atomix-proxy log:

`2023-02-09T06:05:47.825Z        INFO    [github.com/atomix/runtime/proxy/pkg/proxy](http://github.com/atomix/runtime/proxy/pkg/proxy)       proxy/runtime.go:138    Reconfigured connection to store        {"Name": "onos-consensus-store", "Namespace": "aether-roc"}
2023-02-09T06:08:49.176Z        WARN    [github.com/atomix/runtime/primitives/pkg/indexedmap/v1](http://github.com/atomix/runtime/primitives/pkg/indexedmap/v1)  v1/proxy.go:676 Events  {"EventsRequest": "id:<name:\"transactions\" > key:\"uuid:310d6b62-2a88-419e-acc0-f2121eccff5e\" ", "error": "rpc error: code = Canceled desc = context canceled"}
2023-02-09T09:33:32.728Z        WARN    [github.com/atomix/runtime/primitives/pkg/indexedmap/v1](http://github.com/atomix/runtime/primitives/pkg/indexedmap/v1)  v1/proxy.go:676 Events  {"EventsRequest": "id:<name:\"transactions\" > key:\"uuid:d177e9e1-a403-4c4f-bf47-4e7a01c136f3\" ", "error": "rpc error: code = Canceled desc = context canceled"}`

Transaction snippet:

![image](https://user-images.githubusercontent.com/9526678/218078070-1eeffc1f-132b-4f50-8f00-2b636c9e5023.png)

Onos-Config snippet:

![image](https://user-images.githubusercontent.com/9526678/218078406-7f4e2699-cafd-482e-8a23-ebd054651d8a.png)

![image](https://user-images.githubusercontent.com/9526678/218077932-562b0b54-7038-421b-b37a-d83636694116.png)

### Here are steps to reproduce the issue:

1.  Cloned the `aether-in-a-box`
2. Updated the aether-2.1/roc-values.yaml
 ![image](https://user-images.githubusercontent.com/9526678/218078746-2b9ceea4-b828-44d5-96b6-88b64cf55124.png
3.  Run the command: CHARTS=release-2.1 make roc-4g-models
4. Created PV,
5. ./kubectl-run-with-pvc.sh data-onos-consensus-store-0 
 ```
chmod a+w /pvcs/data-onos-consensus-store-0/
 chmod 777 -R /pvcs/data-onos-consensus-store-0/
```
6. Exit from pvc mounter
7. Opened the Aether-roc-gui
8. done 3 transactions
9. Then run `make roc-clean`
10. Deleted PV
11. Created PV
12. Again run: CHARTS=release-2.1 make roc-4g-models
13. ./kubectl-run-with-pvc.sh data-onos-consensus-store-0 
   ```
  chmod a+w /pvcs/data-onos-consensus-store-0/
     chmod 777 -R /pvcs/data-onos-consensus-store-0/
```
14. Exit from pvc mounter
15. Opened the GUI made transaction
16. Checked the transaction in onos-cli , This issue might not have been seen in the first attempt. We are seeing it randomly after 2-3 attempts of cleaning and redeploying the roc using the above steps.
  
### Helm Chart Version:

- Atomix-runtime-0.1.9
- Onos-operator-0.5.6
- Onos-topo-1.4.5
- Onos-cli-1.3.13
- Onos-config-1.8.5

**Comment by amitinfo2k** [comment]
@adibrastegarnia and @kuujo Can you look into this issue.

**Comment by adibrastegarnia** [comment]
@amitinfo2k 
When a transaction is committed that means changes applied from onos-config side in the stores. So very likely the issue is not the store itself and those warnings can happen in normal situation as well. There could be multiple reasons that the changes are not applied to the target. One of the reasons can be there is no connection to the target for some reasons. If you try to reproduce the issue and provide me the following outputs then I can debug it further: 

1) Use onos-cli and check control relations by running onos topo get relations -v command. Control relations represent connections to the target. If you don't see any that means there is no connection to the target. 

2) Use onos-cli and check onos topo get entities -v. We need to make sure topo entities are there and correct. 

3) Get the log of onos-config when that incident happens and share it here in a file and I will go through it.

**Comment by amitinfo2k** [comment]
Here's the required information

```
 onos-cli-56dcf98946-88b2d:~$ onos topo get relations -v
ID:                 uuid:96fcdfbf-1a4d-4b83-87d6-cea23dc68f29
Kind ID:            controls
Source Entity ID:   gnmi:onos-config-6644467bb5-nrjg9
Target Entity ID:   connectivity-service-v2
Labels:             <None>
Aspects:
   <None>
```

```
onos-cli-56dcf98946-88b2d:~$ onos topo get entities -v
ID: gnmi:onos-config-6644467bb5-nrjg9
Kind ID: onos-config
Labels: <None>
Source Id's: uuid:96fcdfbf-1a4d-4b83-87d6-cea23dc68f29
Target Id's:
Aspects:
- onos.topo.Lease={"expiration":"2023-02-10T17:02:26.723564852Z"}

ID: connectivity-service-v2
Kind ID: aether
Labels: <None>
Source Id's:
Target Id's: uuid:96fcdfbf-1a4d-4b83-87d6-cea23dc68f29
Aspects:
- onos.topo.MastershipState={"term":"1","nodeId":"uuid:96fcdfbf-1a4d-4b83-87d6-cea23dc68f29"}
- onos.topo.Location={"lat":52.515,"lng":13.3885}
- onos.topo.Configurable={"address":"aether-roc-umbrella-sdcore-adapter-v2-0:5150","type":"aether","version":"2.0.x"}
- onos.topo.Asset={"name":"5G Core"}
- onos.topo.TLSOptions={"insecure":true}
```

onos config logs:

[onos-config-atomix-proxy.log](https://github.com/onosproject/onos-config/files/10709675/onos-config-atomix-proxy.log)
[onos-config.log](https://github.com/onosproject/onos-config/files/10709677/onos-config.log)

**Comment by adibrastegarnia** [comment]
The log you shared with me is immediately after the incident or there is a chance that onos-config restarted?

**Comment by amitinfo2k** [comment]
@adibrastegarnia there is no restart of any components 

```
root@opencord2:/tmp# kubectl get pods -n aether-roc
NAME                                                              READY   STATUS    RESTARTS   AGE
aether-roc-api-76f88f7d66-twgdj                                   1/1     Running   0          35h
aether-roc-gui-v2-7f4467c496-xnqkv                                1/1     Running   0          35h
aether-roc-umbrella-grafana-6d8568b6f6-jpfwb                      2/2     Running   0          35h
aether-roc-umbrella-kube-state-metrics-54b8cf69bc-5qsjf           1/1     Running   0          35h
aether-roc-umbrella-prometheus-acc-alertmanager-67cb9fc9f956j9w   2/2     Running   0          35h
aether-roc-umbrella-prometheus-acc-server-86c47649c4-kqxs5        2/2     Running   0          35h
aether-roc-umbrella-prometheus-amp-server-65998cc7f5-lnmjm        2/2     Running   0          35h
aether-roc-umbrella-sdcore-adapter-v2-0-697c644f95-62jz8          1/1     Running   0          35h
aether-roc-websocket-568db578d4-7q775                             1/1     Running   0          35h
onos-cli-56dcf98946-88b2d                                         1/1     Running   0          35h
onos-config-6644467bb5-nrjg9                                      5/5     Running   0          35h
onos-consensus-store-0                                            1/1     Running   0          35h
onos-topo-6d65b974b5-8wds9                                        2/2     Running   0          35h
prom-label-proxy-acc-86869b85d4-njcgv                             1/1     Running   0          35h
prom-label-proxy-amp-5fcb7d7477-ww8cc                             1/1     Running   0          35h
subscriber-proxy-745769bb4d-hnd2z                                 1/1     Running   0          35h
```

**Comment by adibrastegarnia** [comment]
Actually, would be good to share the output of onos-cli for entities and relations for the case that it works (I mean you said it happened after multiple runs) if you can . I would like to see the mastership state in that case.

**Comment by amitinfo2k** [comment]
Currently we don't have clean setup. We will need to create one. But, as I remember in both the cases for entities and relations o/p was same. We have shown this issue to Sean before filing it. he is also suspecting some connection issue in atomix logs.

**Comment by praj527** [comment]
@adibrastegarnia
Working setup output for entities, relations and transactions.
<img width="524" alt="entity" src="https://user-images.githubusercontent.com/67963061/218251745-67f75d63-28a3-47d4-8815-250524258e9b.png">
<img width="422" alt="relations" src="https://user-images.githubusercontent.com/67963061/218251750-58705563-7c63-434b-9cba-5e647ade3785.png">
<img width="943" alt="transactions" src="https://user-images.githubusercontent.com/67963061/218251754-b82300cc-4f78-441d-a1c7-e779b96bb3bd.png">

**Comment by adibrastegarnia** [comment]
@amitinfo2k 
The issue that I was seeing earlier in the log is related to this and I was thinking about it line https://github.com/onosproject/onos-config/blob/master/pkg/controller/proposal/controller.go#L569. For some reasons the mastership term in the configuration is not greater than current mastership term in the entity of topo. In normal scenario when there is a mastership change then it means a restarts happened but for some reasons mastership term in topo is not reflecting that anymore so it should wait to catch up but it never happens. Kind of we are back in time which is problematic.

**Comment by amitinfo2k** [comment]
@adibrastegarnia Any workaround where it can be resolved by restarting a component ? until we get the fixes.

**Comment by adibrastegarnia** [comment]
@amitinfo2k  If you can please extend the log in line 569 that I sent you above and include mastership term and mastership term in the config status and try to get it to the situation that transaction doesn't apply. 
Technically, if you restart the target, every time that restart happens mastership term in topo gets increased by one so that way it is gonna catch up but the biggest question for me why we are back in time. It looks like some of the data gets completely lost in the topo (last mastership state) and because of that we are back in time.

**Comment by praj527** [comment]
@adibrastegarnia after the restarting `sdcore-adapter` manually we observed that the `term` in the `MastershipState` increased by 1.
<img width="760" alt="image" src="https://user-images.githubusercontent.com/67963061/218994773-5e601d3c-3d8f-48c4-9ba6-0f9692e5dba8.png">

And the newer transactions are getting into `APPLIED` state but the older transactions which was in `COMMITED` state still continues to be in `COMMITED` state.
<img width="941" alt="ticket" src="https://user-images.githubusercontent.com/67963061/218995958-fae4b193-1f06-4bfb-8630-d725b7d494dd.png">


Is there any way we can get those transactions to be applied.

**Comment by kuujo** [comment]
Were the topo entities recreated after restarting everything? I'm confused about where the mastership term went. If the mastership was lost because those entities were just recreated, there may be a simple fix for this.

**Comment by kuujo** [comment]
FYI in reference to the errors in the Atomix store/proxy, they're not errors at all. Atomix is really careful to log errors with the `ERROR` level. Those are `WARN` level messages, which normally indicates an issue that could be problematic if it recurs continuously. In this case, these are just the errors that occur when streams are closed by a client (e.g. watching the store), so they're not indicative of anything.

**Comment by kuujo** [comment]
This should be fixed by #1642. Please reopen if you run into this issue again.

**Comment by praj527** [comment]
We are getting crash with the latest image build from this commit

We used the Aether-in-a-box, deployed using latest and we update the image of onos-config to latest which we build from this merged commit.

Helm Chart Versions are as:
aether-roc-umbrella-2.1.35
atomix-runtime-0.1.9
onos-operator-0.5.6

Please let us know, if we missed anything here.

Crash Logs:

```
# kubectl logs onos-config-85fd86ddf-ml5dx onos-config -n aether-roc -f
2023-04-10T11:01:55.156Z        INFO    main    onos-config/onos-config.go:70   Starting onos-config    {"CAPath": "/etc/onos/certs/tls.cacrt", "KeyPath
": "/etc/onos/certs/tls.key", "CertPath": "/etc/onos/certs/tls.crt", "GRPCPort": 5150, "TopoAddress": "onos-topo:5150", "Plugins": ["localhost:5152", "l
ocalhost:5153"]}
2023-04-10T11:01:55.156Z        INFO    manager manager/manager.go:57   Creating Manager
2023-04-10T11:01:55.156Z        INFO    manager manager/manager.go:67   Starting Manager
2023-04-10T11:01:55.158Z        FATAL   manager manager/manager.go:70   Unable to run Managerunknown service atomix.runtime.indexedmap.v1.IndexedMaps
github.com/onosproject/onos-config/pkg/manager.(*Manager).Run
        /build/pkg/manager/manager.go:70
main.runRootCommand
        /build/cmd/onos-config/onos-config.go:90
github.com/spf13/cobra.(*Command).execute
        /build/vendor/github.com/spf13/cobra/command.go:856
github.com/spf13/cobra.(*Command).ExecuteC
        /build/vendor/github.com/spf13/cobra/command.go:974
github.com/spf13/cobra.(*Command).Execute
        /build/vendor/github.com/spf13/cobra/command.go:902
main.main
        /build/cmd/onos-config/onos-config.go:41
runtime.main
        /usr/local/go/src/runtime/proc.go:250
```
